docs/design/2022-12-07-keyspace.md

```diff
@@ -0,0 +1,200 @@
+
+## Architecture
+
+In this section, we will outline the architecture when utilizing keyspaces.
+
+![keyspace-arch.png](imgs/keyspace-arch.png)
+
+### PD
+
+PD is responsible for managing `Keyspace` metadata. It provides HTTP and RPC interfaces to create, load, and modify `Keyspace` metadata. All other components retrieve the keyspace metadata from PD.
+
+### TiDB
+
+After the keyspace is created on PD, it can be configured on multiple TiDB instances, forming a TiDB cluster for that keyspace. However, a TiDB instance can serve only one keyspace at a specific moment, unless it is restarted and configured to use a different keyspace.
+
+### Client-go
+
+In TiDB, client-go utilizes codec V2 to add a keyspace prefix when the keyspace is set. A client configured with a keyspace can only read and write data from the specified keyspace.
+
+### TiKV
+
+In TiKV, different keyspaces are isolated by different data key prefixes.
+
+### TiFlash
+
+In TiFlash, requests carry the keyspace prefix, and the data structures and storage paths are isolated by the keyspace ID.
+
+## Usage
+
+To use TiDB with `Keyspace`, the following configuration is required:
+1. TiKV needs to [enable API V2](https://github.com/tikv/rfcs/blob/master/text/0069-api-v2.md#tikv-server).
+2. The TiDB configuration should set `keyspace-name` to specify which keyspace is currently served by the TiDB instance.
+3. BR, Lightning, Dumpling, and CDC specify the keyspace by setting `keyspace-name` on the command line.
+
+## Implementation
+
+### PD
+
+#### PD Functionality
+
+PD provides an HTTP interface and an RPC interface to create, load, change the state, and configure keyspace metadata, and is responsible for interacting with the keyspace metadata store. Users can request the `create keyspace` interface by specifying the keyspace name, which generates the keyspace metadata.
+
+#### Keyspace Name
+
+**Keyspace Name** is specified by the creator during the creation of the keyspace, and once the keyspace is created, the keyspace name cannot be bound to another keyspace.
+
+#### Keyspace ID Allocation
+
+**Keyspace ID** is allocated by PD createKeyspace interface. The keyspace id is assigned by auto-incrementing the id, so keyspace ids are not re-assigned. The max keyspace id is 16777216, no new keyspaces can be created after the maximum keyspace id has been reached.
+
+#### Keyspace Metadata
+
+[**Keyspace Meta**](https://github.com/pingcap/kvproto/blob/d9297553c9009f569eaf4350f68a908f7811ee55/proto/keyspacepb.proto#L26C1-L33C2) is built when a keyspace is created, and the keyspace management interface operates on this metadata. The keyspace state in the metadata changes according to the keyspace state machine. The state change logic is as follows:
+![keyspace-state-machine.png](imgs%2Fkeyspace-state-machine.png)
+
+### TiDB
```

Comment:
It would be great if you can describe the metrics & Grafana things here.

Reply:
Thanks, I've added the relevant descriptions below. PTAL~